### PR TITLE
expression: add baseFunctionClass

### PR DESCRIPTION
**Pull request by hanfei1991**
add base function struct. refactor coalese builtin function.
@shenli @coocood @zimulala @tiancaiamao PTAL

### CODE REVIEW

File: expression/builtin.go [hunk 1]
```diff
@@ -100,6 +100,20 @@ type builtinFunc interface {
 	getCtx() context.Context
 }
 
+type baseFunctionClass struct {
```
Comment (shenli) [inline expression/builtin.go]: Add comment for this struct.

File: expression/builtin.go [hunk 2]
```diff
@@ -100,6 +100,20 @@ type builtinFunc interface {
...
+	funcName string
+	minArgs  int
+	maxArgs  int
+}
+
+func (b *baseFunctionClass) checkValid(args []Expression) error {
```
Comment (shenli) [inline expression/builtin.go]: How about rename it to verifyArgs

File: expression/builtin.go [hunk 3]
```diff
@@ -285,6 +299,26 @@ var DynamicFuncs = map[string]int{
 	ast.Values:       0,
 }
 
+type coalesceFunctionClass struct {
```
Comment (shenli) [inline expression/builtin.go]: Add comment for this struct.

**Comment by shenli** [comment]
@coocood PTAL

**Comment by shenli** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**